From 27b935cef9b86cb71b5b15dd86a3f156b1b10f41 Mon Sep 17 00:00:00 2001
From: w2ppx <w2ppx@users.noreply.github.com>
Date: Thu, 4 Jun 2026 15:26:27 +0300
Subject: [PATCH] feat(splithttp): custom session ID generation

Port of XTLS/Xray-core#6258 (Custom sessionID).

Adds two splithttp config fields:
- sessionIDTable: character set ('base62', 'hex', 'HEX', or literal ASCII)
- sessionIDLength: '<n>' or '<from>-<to>' random length range
---
 infra/conf/v4/transport_internet.go       | 11 +++++++++
 transport/internet/splithttp/config.go    | 30 +++++++++++++++++++++++
 transport/internet/splithttp/config.pb.go | 29 ++++++++++++++++++++--
 transport/internet/splithttp/config.proto |  2 ++
 transport/internet/splithttp/dialer.go    |  4 +--
 5 files changed, 71 insertions(+), 5 deletions(-)

diff --git a/infra/conf/v4/transport_internet.go b/infra/conf/v4/transport_internet.go
index 551e589..d241fbf 100644
--- a/infra/conf/v4/transport_internet.go
+++ b/infra/conf/v4/transport_internet.go
@@ -424,6 +424,8 @@ type SplitHTTPConfig struct {
 	ScMaxEachPostBytes   string               `json:"scMaxEachPostBytes"`
 	ScMinPostsIntervalMs string               `json:"scMinPostsIntervalMs"`
 	ScMaxBufferedPosts   int64                `json:"scMaxConcurrentPosts"`
+	SessionIDTable       string               `json:"sessionIDTable"`
+	SessionIDLength      string               `json:"sessionIDLength"`
 	Xmux                 *XmuxConfig          `json:"xmux"`
 	DownloadSettings     *XHTTPDownloadConfig `json:"downloadSettings"`
 	UseBrowserForwarding bool                 `json:"useBrowserForwarding"`
@@ -564,8 +566,17 @@ func (c *SplitHTTPConfig) Build() (proto.Message, error) {
 		ScMaxEachPostBytes:   c.ScMaxEachPostBytes,
 		ScMinPostsIntervalMs: c.ScMinPostsIntervalMs,
 		ScMaxBufferedPosts:   c.ScMaxBufferedPosts,
+		SessionIDTable:       c.SessionIDTable,
+		SessionIDLength:      c.SessionIDLength,
 		UseBrowserForwarding: c.UseBrowserForwarding,
 	}
+	if c.SessionIDTable != "" {
+		for i := 0; i < len(c.SessionIDTable); i++ {
+			if c.SessionIDTable[i] >= 0x80 {
+				return nil, newError("sessionIDTable must contain only ASCII characters")
+			}
+		}
+	}
 	if config.XPaddingKey == "" {
 		config.XPaddingKey = "x_padding"
 	}
diff --git a/transport/internet/splithttp/config.go b/transport/internet/splithttp/config.go
index a737898..775a9db 100644
--- a/transport/internet/splithttp/config.go
+++ b/transport/internet/splithttp/config.go
@@ -6,12 +6,14 @@ import (
 	"fmt"
 	"io"
 	"math/big"
+	mrand "math/rand/v2"
 	"net/http"
 	"strconv"
 	"strings"
 
 	"github.com/exclavenetwork/exclave-core/v5/common/buf"
 	"github.com/exclavenetwork/exclave-core/v5/common/serial"
+	"github.com/exclavenetwork/exclave-core/v5/common/uuid"
 	"github.com/exclavenetwork/exclave-core/v5/transport/internet"
 )
 
@@ -450,6 +452,34 @@ func (c *Config) ExtractMetaFromRequest(req *http.Request, path string) (session
 	return sessionId, seqStr
 }
 
+var (
+	sessionIDBase62Table   = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
+	sessionIDHexTable      = "0123456789abcdef"
+	sessionIDHexUpperTable = "0123456789ABCDEF"
+)
+
+func (c *Config) GenerateSessionID() string {
+	length := newRandRangeConfig(0, 0, c.SessionIDLength).rand()
+	table := c.SessionIDTable
+	switch table {
+	case "base62":
+		table = sessionIDBase62Table
+	case "hex":
+		table = sessionIDHexTable
+	case "HEX":
+		table = sessionIDHexUpperTable
+	}
+	if table != "" && length > 0 {
+		id := make([]byte, length)
+		for i := range id {
+			id[i] = table[mrand.IntN(len(table))]
+		}
+		return string(id)
+	}
+	u := uuid.New()
+	return u.String()
+}
+
 func appendToPath(path, value string) string {
 	if strings.HasSuffix(path, "/") {
 		return path + value
diff --git a/transport/internet/splithttp/config.pb.go b/transport/internet/splithttp/config.pb.go
index cb97e30..3c25f92 100644
--- a/transport/internet/splithttp/config.pb.go
+++ b/transport/internet/splithttp/config.pb.go
@@ -1,3 +1,9 @@
+// Code generated by protoc-gen-go. DO NOT EDIT.
+// versions:
+// 	protoc-gen-go v1.36.11
+// 	protoc        v5.28.3
+// source: transport/internet/splithttp/config.proto
+
 package splithttp
 
 import (
@@ -205,6 +211,8 @@ type Config struct {
 	UplinkDataPlacement  string                 `protobuf:"bytes,22,opt,name=uplinkDataPlacement,proto3" json:"uplinkDataPlacement,omitempty"`
 	UplinkDataKey        string                 `protobuf:"bytes,23,opt,name=uplinkDataKey,proto3" json:"uplinkDataKey,omitempty"`
 	UplinkChunkSize      string                 `protobuf:"bytes,24,opt,name=uplinkChunkSize,proto3" json:"uplinkChunkSize,omitempty"`
+	SessionIDTable       string                 `protobuf:"bytes,25,opt,name=sessionIDTable,proto3" json:"sessionIDTable,omitempty"`
+	SessionIDLength      string                 `protobuf:"bytes,26,opt,name=sessionIDLength,proto3" json:"sessionIDLength,omitempty"`
 	UseBrowserForwarding bool                   `protobuf:"varint,99,opt,name=use_browser_forwarding,json=useBrowserForwarding,proto3" json:"use_browser_forwarding,omitempty"`
 	unknownFields        protoimpl.UnknownFields
 	sizeCache            protoimpl.SizeCache
@@ -408,6 +416,20 @@ func (x *Config) GetUplinkChunkSize() string {
 	return ""
 }
 
+func (x *Config) GetSessionIDTable() string {
+	if x != nil {
+		return x.SessionIDTable
+	}
+	return ""
+}
+
+func (x *Config) GetSessionIDLength() string {
+	if x != nil {
+		return x.SessionIDLength
+	}
+	return ""
+}
+
 func (x *Config) GetUseBrowserForwarding() bool {
 	if x != nil {
 		return x.UseBrowserForwarding
@@ -433,7 +455,8 @@ const file_transport_internet_splithttp_config_proto_rawDesc = "" +
 	"\x12transport_settings\x18\x03 \x01(\v2\x14.google.protobuf.AnyR\x11transportSettings\x12#\n" +
 	"\rsecurity_type\x18\x04 \x01(\tR\fsecurityType\x12A\n" +
 	"\x11security_settings\x18\x05 \x01(\v2\x14.google.protobuf.AnyR\x10securitySettings\x12V\n" +
-	"\x0fsocket_settings\x18\x06 \x01(\v2-.exclave.core.transport.internet.SocketConfigR\x0esocketSettings\"\xe6\t\n" +
+	"\x0fsocket_settings\x18\x06 \x01(\v2-.exclave.core.transport.internet.SocketConfigR\x0esocketSettings\"\xb8\n" +
+	"\n" +
 	"\x06Config\x12\x12\n" +
 	"\x04host\x18\x01 \x01(\tR\x04host\x12\x12\n" +
 	"\x04path\x18\x02 \x01(\tR\x04path\x12\x12\n" +
@@ -461,7 +484,9 @@ const file_transport_internet_splithttp_config_proto_rawDesc = "" +
 	"\x06seqKey\x18\x15 \x01(\tR\x06seqKey\x120\n" +
 	"\x13uplinkDataPlacement\x18\x16 \x01(\tR\x13uplinkDataPlacement\x12$\n" +
 	"\ruplinkDataKey\x18\x17 \x01(\tR\ruplinkDataKey\x12(\n" +
-	"\x0fuplinkChunkSize\x18\x18 \x01(\tR\x0fuplinkChunkSize\x124\n" +
+	"\x0fuplinkChunkSize\x18\x18 \x01(\tR\x0fuplinkChunkSize\x12&\n" +
+	"\x0esessionIDTable\x18\x19 \x01(\tR\x0esessionIDTable\x12(\n" +
+	"\x0fsessionIDLength\x18\x1a \x01(\tR\x0fsessionIDLength\x124\n" +
 	"\x16use_browser_forwarding\x18c \x01(\bR\x14useBrowserForwarding\x1a:\n" +
 	"\fHeadersEntry\x12\x10\n" +
 	"\x03key\x18\x01 \x01(\tR\x03key\x12\x14\n" +
diff --git a/transport/internet/splithttp/config.proto b/transport/internet/splithttp/config.proto
index 974e474..3032a26 100644
--- a/transport/internet/splithttp/config.proto
+++ b/transport/internet/splithttp/config.proto
@@ -56,5 +56,7 @@ message Config {
   string uplinkDataPlacement = 22;
   string uplinkDataKey = 23;
   string uplinkChunkSize = 24;
+  string sessionIDTable = 25;
+  string sessionIDLength = 26;
   bool use_browser_forwarding = 99;
 }
diff --git a/transport/internet/splithttp/dialer.go b/transport/internet/splithttp/dialer.go
index ec48510..921ebdf 100644
--- a/transport/internet/splithttp/dialer.go
+++ b/transport/internet/splithttp/dialer.go
@@ -25,7 +25,6 @@ import (
 	"github.com/exclavenetwork/exclave-core/v5/common/net"
 	"github.com/exclavenetwork/exclave-core/v5/common/session"
 	"github.com/exclavenetwork/exclave-core/v5/common/signal/done"
-	"github.com/exclavenetwork/exclave-core/v5/common/uuid"
 	"github.com/exclavenetwork/exclave-core/v5/features/extension"
 	"github.com/exclavenetwork/exclave-core/v5/transport/internet"
 	"github.com/exclavenetwork/exclave-core/v5/transport/internet/reality"
@@ -306,8 +305,7 @@ func Dial(ctx context.Context, dest net.Destination, streamSettings *internet.Me
 
 	sessionId := ""
 	if mode != "stream-one" {
-		sessionIdUuid := uuid.New()
-		sessionId = sessionIdUuid.String()
+		sessionId = transportConfiguration.GenerateSessionID()
 	}
 
 	newError(fmt.Sprintf("XHTTP is dialing to %s, mode %s, HTTP version %s, host %s", dest, mode, httpVersion, requestURL.Host)).AtInfo().WriteToLog(session.ExportIDToError(ctx))
-- 
2.43.0

